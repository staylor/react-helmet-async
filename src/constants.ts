export enum TAG_PROPERTIES {
  CHARSET = 'charset',
  CSS_TEXT = 'cssText',
  HREF = 'href',
  HTTPEQUIV = 'http-equiv',
  INNER_HTML = 'innerHTML',
  ITEM_PROP = 'itemprop',
  NAME = 'name',
  PROPERTY = 'property',
  REL = 'rel',
  SRC = 'src',
}

export enum ATTRIBUTE_NAMES {
  BODY = 'bodyAttributes',
  HTML = 'htmlAttributes',
  TITLE = 'titleAttributes',
}

export enum TAG_NAMES {
  BASE = 'base',
  BODY = 'body',
  HEAD = 'head',
  HTML = 'html',
  LINK = 'link',
  META = 'meta',
  NOSCRIPT = 'noscript',
  SCRIPT = 'script',
  STYLE = 'style',
  TITLE = 'title',
  FRAGMENT = 'Symbol(react.fragment)',
}

export const SEO_PRIORITY_TAGS = {
  link: { rel: ['amphtml', 'canonical', 'alternate'] },
  script: { type: ['application/ld+json'] },
  meta: {
    charset: '',
    name: ['generator', 'robots', 'description'],
    property: [
      'og:type',
      'og:title',
      'og:url',
      'og:image',
      'og:image:alt',
      'og:description',
      'twitter:url',
      'twitter:title',
      'twitter:description',
      'twitter:image',
      'twitter:image:alt',
      'twitter:card',
      'twitter:site',
    ],
  },
};

export const VALID_TAG_NAMES = Object.values(TAG_NAMES);

export const REACT_TAG_MAP = {
  accesskey: 'accessKey',
  charset: 'charSet',
  class: 'className',
  contenteditable: 'contentEditable',
  contextmenu: 'contextMenu',
  'http-equiv': 'httpEquiv',
  itemprop: 'itemProp',
  tabindex: 'tabIndex',
};

export const HTML_TAG_MAP = Object.entries(REACT_TAG_MAP).reduce(
  (carry, [key, value]) => {
    carry[value] = key;
    return carry;
  },
  {} as { [key: string]: string }
);

export const HELMET_ATTRIBUTE = 'data-rh';
