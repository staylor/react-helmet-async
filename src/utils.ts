import { TAG_NAMES, TAG_PROPERTIES, ATTRIBUTE_NAMES } from './constants';

interface PropList {
  [key: string]: any;
}

type PropsList = PropList[];
type AttributeList = string[];

interface SeenTags {
  [key: string]: {
    [key: string]: boolean;
  };
}

interface MatchProps {
  [key: string]: string | AttributeList;
}

const HELMET_PROPS = {
  DEFAULT_TITLE: 'defaultTitle',
  DEFER: 'defer',
  ENCODE_SPECIAL_CHARACTERS: 'encodeSpecialCharacters',
  ON_CHANGE_CLIENT_STATE: 'onChangeClientState',
  TITLE_TEMPLATE: 'titleTemplate',
  PRIORITIZE_SEO_TAGS: 'prioritizeSeoTags',
};

const getInnermostProperty = (propsList: PropsList, property: string) => {
  for (let i = propsList.length - 1; i >= 0; i -= 1) {
    const props = propsList[i];

    if (Object.prototype.hasOwnProperty.call(props, property)) {
      return props[property];
    }
  }

  return null;
};

const getTitleFromPropsList = (propsList: PropsList) => {
  let innermostTitle = getInnermostProperty(propsList, TAG_NAMES.TITLE);
  const innermostTemplate = getInnermostProperty(propsList, HELMET_PROPS.TITLE_TEMPLATE);
  if (Array.isArray(innermostTitle)) {
    innermostTitle = innermostTitle.join('');
  }
  if (innermostTemplate && innermostTitle) {
    // use function arg to avoid need to escape $ characters
    return innermostTemplate.replace(/%s/g, () => innermostTitle);
  }

  const innermostDefaultTitle = getInnermostProperty(propsList, HELMET_PROPS.DEFAULT_TITLE);

  return innermostTitle || innermostDefaultTitle || undefined;
};

const getOnChangeClientState = (propsList: PropsList) =>
  getInnermostProperty(propsList, HELMET_PROPS.ON_CHANGE_CLIENT_STATE) || (() => {});

const getAttributesFromPropsList = (tagType: string, propsList: PropsList) =>
  propsList
    .filter(props => typeof props[tagType] !== 'undefined')
    .map(props => props[tagType])
    .reduce((tagAttrs, current) => ({ ...tagAttrs, ...current }), {});

const getBaseTagFromPropsList = (primaryAttributes: AttributeList, propsList: PropsList) =>
  propsList
    .filter(props => typeof props[TAG_NAMES.BASE] !== 'undefined')
    .map(props => props[TAG_NAMES.BASE])
    .reverse()
    .reduce((innermostBaseTag, tag) => {
      if (!innermostBaseTag.length) {
        const keys = Object.keys(tag);

        for (let i = 0; i < keys.length; i += 1) {
          const attributeKey = keys[i];
          const lowerCaseAttributeKey = attributeKey.toLowerCase();

          if (
            primaryAttributes.indexOf(lowerCaseAttributeKey) !== -1 &&
            tag[lowerCaseAttributeKey]
          ) {
            return innermostBaseTag.concat(tag);
          }
        }
      }

      return innermostBaseTag;
    }, []);

const warn = (msg: string) => console && typeof console.warn === 'function' && console.warn(msg);

const getTagsFromPropsList = (
  tagName: string,
  primaryAttributes: AttributeList,
  propsList: PropsList
) => {
  // Calculate list of tags, giving priority innermost component (end of the propslist)
  const approvedSeenTags: SeenTags = {};

  return propsList
    .filter(props => {
      if (Array.isArray(props[tagName])) {
        return true;
      }
      if (typeof props[tagName] !== 'undefined') {
        warn(
          `Helmet: ${tagName} should be of type "Array". Instead found type "${typeof props[
            tagName
          ]}"`
        );
      }
      return false;
    })
    .map(props => props[tagName])
    .reverse()
    .reduce((approvedTags, instanceTags) => {
      const instanceSeenTags: SeenTags = {};

      instanceTags
        .filter((tag: PropList) => {
          let primaryAttributeKey;
          const keys = Object.keys(tag);
          for (let i = 0; i < keys.length; i += 1) {
            const attributeKey = keys[i];
            const lowerCaseAttributeKey = attributeKey.toLowerCase();

            // Special rule with link tags, since rel and href are both primary tags, rel takes priority
            if (
              primaryAttributes.indexOf(lowerCaseAttributeKey) !== -1 &&
              !(
                primaryAttributeKey === TAG_PROPERTIES.REL &&
                tag[primaryAttributeKey].toLowerCase() === 'canonical'
              ) &&
              !(
                lowerCaseAttributeKey === TAG_PROPERTIES.REL &&
                tag[lowerCaseAttributeKey].toLowerCase() === 'stylesheet'
              )
            ) {
              primaryAttributeKey = lowerCaseAttributeKey;
            }
            // Special case for innerHTML which doesn't work lowercased
            if (
              primaryAttributes.indexOf(attributeKey) !== -1 &&
              (attributeKey === TAG_PROPERTIES.INNER_HTML ||
                attributeKey === TAG_PROPERTIES.CSS_TEXT ||
                attributeKey === TAG_PROPERTIES.ITEM_PROP)
            ) {
              primaryAttributeKey = attributeKey;
            }
          }

          if (!primaryAttributeKey || !tag[primaryAttributeKey]) {
            return false;
          }

          const value = tag[primaryAttributeKey].toLowerCase();

          if (!approvedSeenTags[primaryAttributeKey]) {
            approvedSeenTags[primaryAttributeKey] = {};
          }

          if (!instanceSeenTags[primaryAttributeKey]) {
            instanceSeenTags[primaryAttributeKey] = {};
          }

          if (!approvedSeenTags[primaryAttributeKey][value]) {
            instanceSeenTags[primaryAttributeKey][value] = true;
            return true;
          }

          return false;
        })
        .reverse()
        .forEach((tag: PropList) => approvedTags.push(tag));

      // Update seen tags with tags from this instance
      const keys = Object.keys(instanceSeenTags);
      for (let i = 0; i < keys.length; i += 1) {
        const attributeKey = keys[i];
        const tagUnion = {
          ...approvedSeenTags[attributeKey],
          ...instanceSeenTags[attributeKey],
        };

        approvedSeenTags[attributeKey] = tagUnion;
      }

      return approvedTags;
    }, [])
    .reverse();
};

const getAnyTrueFromPropsList = (propsList: PropsList, checkedTag: string) => {
  if (Array.isArray(propsList) && propsList.length) {
    for (let index = 0; index < propsList.length; index += 1) {
      const prop = propsList[index];
      if (prop[checkedTag]) {
        return true;
      }
    }
  }
  return false;
};

const reducePropsToState = (propsList: PropsList) => ({
  baseTag: getBaseTagFromPropsList([TAG_PROPERTIES.HREF], propsList),
  bodyAttributes: getAttributesFromPropsList(ATTRIBUTE_NAMES.BODY, propsList),
  defer: getInnermostProperty(propsList, HELMET_PROPS.DEFER),
  encode: getInnermostProperty(propsList, HELMET_PROPS.ENCODE_SPECIAL_CHARACTERS),
  htmlAttributes: getAttributesFromPropsList(ATTRIBUTE_NAMES.HTML, propsList),
  linkTags: getTagsFromPropsList(
    TAG_NAMES.LINK,
    [TAG_PROPERTIES.REL, TAG_PROPERTIES.HREF],
    propsList
  ),
  metaTags: getTagsFromPropsList(
    TAG_NAMES.META,
    [
      TAG_PROPERTIES.NAME,
      TAG_PROPERTIES.CHARSET,
      TAG_PROPERTIES.HTTPEQUIV,
      TAG_PROPERTIES.PROPERTY,
      TAG_PROPERTIES.ITEM_PROP,
    ],
    propsList
  ),
  noscriptTags: getTagsFromPropsList(TAG_NAMES.NOSCRIPT, [TAG_PROPERTIES.INNER_HTML], propsList),
  onChangeClientState: getOnChangeClientState(propsList),
  scriptTags: getTagsFromPropsList(
    TAG_NAMES.SCRIPT,
    [TAG_PROPERTIES.SRC, TAG_PROPERTIES.INNER_HTML],
    propsList
  ),
  styleTags: getTagsFromPropsList(TAG_NAMES.STYLE, [TAG_PROPERTIES.CSS_TEXT], propsList),
  title: getTitleFromPropsList(propsList),
  titleAttributes: getAttributesFromPropsList(ATTRIBUTE_NAMES.TITLE, propsList),
  prioritizeSeoTags: getAnyTrueFromPropsList(propsList, HELMET_PROPS.PRIORITIZE_SEO_TAGS),
});

export const flattenArray = (possibleArray: string[] | string) =>
  Array.isArray(possibleArray) ? possibleArray.join('') : possibleArray;

export { reducePropsToState };

const checkIfPropsMatch = (props: PropList, toMatch: MatchProps) => {
  const keys = Object.keys(props);
  for (let i = 0; i < keys.length; i += 1) {
    // e.g. if rel exists in the list of allowed props [amphtml, alternate, etc]
    if (toMatch[keys[i]] && toMatch[keys[i]].includes(props[keys[i]])) {
      return true;
    }
  }
  return false;
};

export const prioritizer = (elementsList: HTMLElement[], propsToMatch: MatchProps) => {
  if (Array.isArray(elementsList)) {
    return elementsList.reduce(
      (acc, elementAttrs) => {
        if (checkIfPropsMatch(elementAttrs, propsToMatch)) {
          acc.priority.push(elementAttrs);
        } else {
          acc.default.push(elementAttrs);
        }
        return acc;
      },
      { priority: [] as HTMLElement[], default: [] as HTMLElement[] }
    );
  }
  return { default: elementsList, priority: [] };
};

export const without = (obj: PropList, key: string) => {
  return {
    ...obj,
    [key]: undefined,
  };
};
