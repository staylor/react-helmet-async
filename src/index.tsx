import type { PropsWithChildren, ReactElement, ReactNode } from 'react';
import React, { Component } from 'react';
import fastCompare from 'react-fast-compare';
import invariant from 'invariant';

import { Context } from './Provider';
import type { HelmetDataType } from './HelmetData';
import HelmetData from './HelmetData';
import type { DispatcherContextProp } from './Dispatcher';
import Dispatcher from './Dispatcher';
import { without } from './utils';
import { TAG_NAMES, VALID_TAG_NAMES, HTML_TAG_MAP } from './constants';
import type { HelmetProps } from './types';

export * from './types';

export { default as HelmetData } from './HelmetData';
export { default as HelmetProvider } from './Provider';

type Props = { [key: string]: any };

export class Helmet extends Component<PropsWithChildren<HelmetProps>> {
  static defaultProps = {
    defer: true,
    encodeSpecialCharacters: true,
    prioritizeSeoTags: false,
  };

  shouldComponentUpdate(nextProps: HelmetProps) {
    return !fastCompare(without(this.props, 'helmetData'), without(nextProps, 'helmetData'));
  }

  mapNestedChildrenToProps(child: ReactElement, nestedChildren: ReactNode) {
    if (!nestedChildren) {
      return null;
    }

    switch (child.type) {
      case TAG_NAMES.SCRIPT:
      case TAG_NAMES.NOSCRIPT:
        return {
          innerHTML: nestedChildren,
        };

      case TAG_NAMES.STYLE:
        return {
          cssText: nestedChildren,
        };
      default:
        throw new Error(
          `<${child.type} /> elements are self-closing and can not contain children. Refer to our API for more information.`
        );
    }
  }

  flattenArrayTypeChildren(
    child: JSX.Element,
    arrayTypeChildren: { [key: string]: JSX.Element[] },
    newChildProps: Props,
    nestedChildren: ReactNode
  ) {
    return {
      ...arrayTypeChildren,
      [child.type]: [
        ...(arrayTypeChildren[child.type] || []),
        {
          ...newChildProps,
          ...this.mapNestedChildrenToProps(child, nestedChildren),
        },
      ],
    };
  }

  mapObjectTypeChildren(
    child: JSX.Element,
    newProps: Props,
    newChildProps: Props,
    nestedChildren: ReactNode
  ) {
    switch (child.type) {
      case TAG_NAMES.TITLE:
        return {
          ...newProps,
          [child.type]: nestedChildren,
          titleAttributes: { ...newChildProps },
        };

      case TAG_NAMES.BODY:
        return {
          ...newProps,
          bodyAttributes: { ...newChildProps },
        };

      case TAG_NAMES.HTML:
        return {
          ...newProps,
          htmlAttributes: { ...newChildProps },
        };
      default:
        return {
          ...newProps,
          [child.type]: { ...newChildProps },
        };
    }
  }

  mapArrayTypeChildrenToProps(arrayTypeChildren: { [key: string]: JSX.Element }, newProps: Props) {
    let newFlattenedProps = { ...newProps };

    Object.keys(arrayTypeChildren).forEach(arrayChildName => {
      newFlattenedProps = {
        ...newFlattenedProps,
        [arrayChildName]: arrayTypeChildren[arrayChildName],
      };
    });

    return newFlattenedProps;
  }

  warnOnInvalidChildren(child: JSX.Element, nestedChildren: ReactNode) {
    invariant(
      VALID_TAG_NAMES.some(name => child.type === name),
      typeof child.type === 'function'
        ? `You may be attempting to nest <Helmet> components within each other, which is not allowed. Refer to our API for more information.`
        : `Only elements types ${VALID_TAG_NAMES.join(
            ', '
          )} are allowed. Helmet does not support rendering <${
            child.type
          }> elements. Refer to our API for more information.`
    );

    invariant(
      !nestedChildren ||
        typeof nestedChildren === 'string' ||
        (Array.isArray(nestedChildren) &&
          !nestedChildren.some(nestedChild => typeof nestedChild !== 'string')),
      `Helmet expects a string as a child of <${child.type}>. Did you forget to wrap your children in braces? ( <${child.type}>{\`\`}</${child.type}> ) Refer to our API for more information.`
    );

    return true;
  }

  mapChildrenToProps(children: ReactNode, newProps: Props) {
    let arrayTypeChildren = {};

    React.Children.forEach(children as JSX.Element, (child: ReactElement) => {
      if (!child || !child.props) {
        return;
      }

      const { children: nestedChildren, ...childProps } = child.props;
      // convert React props to HTML attributes
      const newChildProps = Object.keys(childProps).reduce((obj: Props, key) => {
        obj[HTML_TAG_MAP[key] || key] = childProps[key];
        return obj;
      }, {});

      let { type } = child;
      if (typeof type === 'symbol') {
        type = (type as 'symbol').toString();
      } else {
        this.warnOnInvalidChildren(child, nestedChildren);
      }

      switch (type) {
        case TAG_NAMES.FRAGMENT:
          newProps = this.mapChildrenToProps(nestedChildren, newProps);
          break;

        case TAG_NAMES.LINK:
        case TAG_NAMES.META:
        case TAG_NAMES.NOSCRIPT:
        case TAG_NAMES.SCRIPT:
        case TAG_NAMES.STYLE:
          arrayTypeChildren = this.flattenArrayTypeChildren(
            child,
            arrayTypeChildren,
            newChildProps,
            nestedChildren
          );
          break;

        default:
          newProps = this.mapObjectTypeChildren(child, newProps, newChildProps, nestedChildren);
          break;
      }
    });

    return this.mapArrayTypeChildrenToProps(arrayTypeChildren, newProps);
  }

  render() {
    const { children, ...props } = this.props;
    let newProps = { ...props };
    let { helmetData } = props;

    if (children) {
      newProps = this.mapChildrenToProps(children, newProps);
    }

    if (helmetData && !(helmetData instanceof HelmetData)) {
      const data = helmetData as HelmetDataType;
      helmetData = new HelmetData(data.context, true);
      delete newProps.helmetData;
    }

    return helmetData ? (
      <Dispatcher {...newProps} context={helmetData.value} />
    ) : (
      <Context.Consumer>
        {context => <Dispatcher {...newProps} context={context as DispatcherContextProp} />}
      </Context.Consumer>
    );
  }
}
