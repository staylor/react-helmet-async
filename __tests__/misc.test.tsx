import React from 'react';

import { Helmet } from '../src';
import { HELMET_ATTRIBUTE } from '../src/constants';

import { render } from './utils';

// TODO: This is confusing
Helmet.defaultProps.defer = false;

describe('misc', () => {
  describe('API', () => {
    it('encodes special characters', () => {
      render(
        <Helmet
          meta={[
            {
              name: 'description',
              content: 'This is "quoted" text and & and \'.',
            },
          ]}
        />
      );

      const existingTags = document.head.querySelectorAll(`meta[${HELMET_ATTRIBUTE}]`);
      const existingTag = existingTags[0];

      expect(existingTags).toBeDefined();
      expect(existingTags).toHaveLength(1);

      expect(existingTag).toBeInstanceOf(Element);
      expect(existingTag.getAttribute).toBeDefined();
      expect(existingTag).toHaveAttribute('name', 'description');
      expect(existingTag).toHaveAttribute('content', 'This is "quoted" text and & and \'.');
      expect(existingTag.outerHTML).toMatchSnapshot();
    });

    it('does not change the DOM if it recevies identical props', () => {
      const onChange = vi.fn();

      render(
        <Helmet
          meta={[{ name: 'description', content: 'Test description' }]}
          title="Test Title"
          onChangeClientState={onChange}
        />
      );

      // Re-rendering will pass new props to an already mounted Helmet
      render(
        <Helmet
          meta={[{ name: 'description', content: 'Test description' }]}
          title="Test Title"
          onChangeClientState={onChange}
        />
      );

      expect(onChange).toBeCalledTimes(1);
    });

    it('does not write the DOM if the client and server are identical', () => {
      document.head.innerHTML = `<script ${HELMET_ATTRIBUTE}="true" src="http://localhost/test.js" type="text/javascript" />`;

      const onChange = vi.fn();
      render(
        <Helmet
          script={[
            {
              src: 'http://localhost/test.js',
              type: 'text/javascript',
            },
          ]}
          onChangeClientState={onChange}
        />
      );

      expect(onChange).toHaveBeenCalled();

      const [, addedTags, removedTags] = onChange.mock.calls[0];

      expect(addedTags).toEqual({});
      expect(removedTags).toEqual({});
    });

    it('only adds new tags and preserves tags when rendering additional Helmet instances', () => {
      const onChange = vi.fn();
      let addedTags;
      let removedTags;
      render(
        <Helmet
          link={[
            {
              href: 'http://localhost/style.css',
              rel: 'stylesheet',
              type: 'text/css',
            },
          ]}
          meta={[{ name: 'description', content: 'Test description' }]}
          onChangeClientState={onChange}
        />
      );

      expect(onChange).toHaveBeenCalled();

      addedTags = onChange.mock.calls[0][1];
      removedTags = onChange.mock.calls[0][2];

      expect(addedTags).toHaveProperty('metaTags');
      expect(addedTags.metaTags[0]).toBeDefined();
      expect(addedTags.metaTags[0].outerHTML).toMatchSnapshot();
      expect(addedTags).toHaveProperty('linkTags');
      expect(addedTags.linkTags[0]).toBeDefined();
      expect(addedTags.linkTags[0].outerHTML).toMatchSnapshot();
      expect(removedTags).toEqual({});

      // Re-rendering will pass new props to an already mounted Helmet
      render(
        <Helmet
          link={[
            {
              href: 'http://localhost/style.css',
              rel: 'stylesheet',
              type: 'text/css',
            },
            {
              href: 'http://localhost/style2.css',
              rel: 'stylesheet',
              type: 'text/css',
            },
          ]}
          meta={[{ name: 'description', content: 'New description' }]}
          onChangeClientState={onChange}
        />
      );

      expect(onChange).toBeCalledTimes(2);

      addedTags = onChange.mock.calls[1][1];
      removedTags = onChange.mock.calls[1][2];

      expect(addedTags).toHaveProperty('metaTags');
      expect(addedTags.metaTags[0]).toBeDefined();
      expect(addedTags.metaTags[0].outerHTML).toMatchSnapshot();
      expect(addedTags).toHaveProperty('linkTags');
      expect(addedTags.linkTags[0]).toBeDefined();
      expect(addedTags.linkTags[0].outerHTML).toMatchSnapshot();
      expect(removedTags).toHaveProperty('metaTags');
      expect(removedTags.metaTags[0]).toBeDefined();
      expect(removedTags.metaTags[0].outerHTML).toMatchSnapshot();
      expect(removedTags).not.toHaveProperty('linkTags');
    });

    it('does not accept nested Helmets', () => {
      const consoleError = global.console.error;
      global.console.error = vi.fn();

      const renderInvalid = () => {
        render(
          <Helmet title="Test Title">
            <Helmet title={"Title you'll never see"} />
          </Helmet>
        );
      };

      expect(renderInvalid).toThrow(
        'You may be attempting to nest <Helmet> components within each other, which is not allowed. Refer to our API for more information.'
      );

      global.console.error = consoleError;
    });

    it('recognizes valid tags regardless of attribute ordering', () => {
      render(<Helmet meta={[{ content: 'Test Description', name: 'description' }]} />);

      const existingTags = document.head.querySelectorAll(`meta[${HELMET_ATTRIBUTE}]`);
      const existingTag = existingTags[0];

      expect(existingTags).toBeDefined();
      expect(existingTags).toHaveLength(1);

      expect(existingTag).toBeInstanceOf(Element);
      expect(existingTag.getAttribute).toBeDefined();
      expect(existingTag).toHaveAttribute('name', 'description');
      expect(existingTag).toHaveAttribute('content', 'Test Description');
      expect(existingTag.outerHTML).toMatchSnapshot();
    });

    it('requestAnimationFrame works as expected', () => {
      return new Promise(resolve => {
        requestAnimationFrame(cb => {
          expect(cb).toBeDefined();
          expect(typeof cb).toBe('number');

          resolve(true);
        });
      });
    });
  });

  describe('Declarative API', () => {
    it('encodes special characters', () => {
      render(
        <Helmet>
          <meta name="description" content={'This is "quoted" text and & and \'.'} />
        </Helmet>
      );

      const existingTags = document.head.querySelectorAll(`meta[${HELMET_ATTRIBUTE}]`);
      const existingTag = existingTags[0];

      expect(existingTags).toBeDefined();
      expect(existingTags).toHaveLength(1);
      expect(existingTag).toBeInstanceOf(Element);
      expect(existingTag.getAttribute).toBeDefined();
      expect(existingTag).toHaveAttribute('name', 'description');
      expect(existingTag).toHaveAttribute('content', 'This is "quoted" text and & and \'.');
      expect(existingTag.outerHTML).toMatchSnapshot();
    });

    it('does not change the DOM if it recevies identical props', () => {
      const onChange = vi.fn();
      render(
        <Helmet onChangeClientState={onChange}>
          <meta name="description" content="Test description" />
          <title>Test Title</title>
        </Helmet>
      );

      // Re-rendering will pass new props to an already mounted Helmet
      render(
        <Helmet onChangeClientState={onChange}>
          <meta name="description" content="Test description" />
          <title>Test Title</title>
        </Helmet>
      );

      expect(onChange).toBeCalledTimes(1);
    });

    it('does not write the DOM if the client and server are identical', () => {
      document.head.innerHTML = `<script ${HELMET_ATTRIBUTE}="true" src="http://localhost/test.js" type="text/javascript" />`;

      const onChange = vi.fn();
      render(
        <Helmet onChangeClientState={onChange}>
          <script src="http://localhost/test.js" type="text/javascript" />
        </Helmet>
      );

      expect(onChange).toHaveBeenCalled();

      const [, addedTags, removedTags] = onChange.mock.calls[0];

      expect(addedTags).toEqual({});
      expect(removedTags).toEqual({});
    });

    it('only adds new tags and preserves tags when rendering additional Helmet instances', () => {
      const onChange = vi.fn();
      let addedTags;
      let removedTags;

      render(
        <Helmet onChangeClientState={onChange}>
          <link href="http://localhost/style.css" rel="stylesheet" type="text/css" />
          <meta name="description" content="Test description" />
        </Helmet>
      );

      expect(onChange).toHaveBeenCalled();

      addedTags = onChange.mock.calls[0][1];
      removedTags = onChange.mock.calls[0][2];

      expect(addedTags).toHaveProperty('metaTags');
      expect(addedTags.metaTags[0]).toBeDefined();
      expect(addedTags.metaTags[0].outerHTML).toMatchSnapshot();
      expect(addedTags).toHaveProperty('linkTags');
      expect(addedTags.linkTags[0]).toBeDefined();
      expect(addedTags.linkTags[0].outerHTML).toMatchSnapshot();
      expect(removedTags).toEqual({});

      // Re-rendering will pass new props to an already mounted Helmet
      render(
        <Helmet onChangeClientState={onChange}>
          <link href="http://localhost/style.css" rel="stylesheet" type="text/css" />
          <link href="http://localhost/style2.css" rel="stylesheet" type="text/css" />
          <meta name="description" content="New description" />
        </Helmet>
      );

      expect(onChange).toBeCalledTimes(2);

      addedTags = onChange.mock.calls[1][1];
      removedTags = onChange.mock.calls[1][2];

      expect(addedTags).toHaveProperty('metaTags');
      expect(addedTags.metaTags[0]).toBeDefined();
      expect(addedTags.metaTags[0].outerHTML).toMatchSnapshot();
      expect(addedTags).toHaveProperty('linkTags');
      expect(addedTags.linkTags[0]).toBeDefined();
      expect(addedTags.linkTags[0].outerHTML).toMatchSnapshot();
      expect(removedTags).toHaveProperty('metaTags');
      expect(removedTags.metaTags[0]).toBeDefined();
      expect(removedTags.metaTags[0].outerHTML).toMatchSnapshot();
      expect(removedTags).not.toHaveProperty('linkTags');
    });

    it('does not accept nested Helmets', () => {
      const consoleError = global.console.error;
      global.console.error = vi.fn();

      const renderInvalid = () => {
        render(
          <Helmet>
            <title>Test Title</title>
            <Helmet>
              <title>Title you will never see</title>
            </Helmet>
          </Helmet>
        );
      };

      expect(renderInvalid).toThrow(
        'You may be attempting to nest <Helmet> components within each other, which is not allowed. Refer to our API for more information.'
      );

      global.console.error = consoleError;
    });

    it('throws on invalid elements', () => {
      const consoleError = global.console.error;
      global.console.error = vi.fn();

      const renderInvalid = () => {
        render(
          <Helmet>
            <title>Test Title</title>
            <div>
              <title>Title you will never see</title>
            </div>
          </Helmet>
        );
      };

      expect(renderInvalid).toThrow(
        'Only elements types base, body, head, html, link, meta, noscript, script, style, title, Symbol(react.fragment) are allowed. Helmet does not support rendering <div> elements. Refer to our API for more information.'
      );

      global.console.error = consoleError;
    });

    it('throws on invalid self-closing elements', () => {
      const consoleError = global.console.error;
      global.console.error = vi.fn();

      const renderInvalid = () => {
        render(
          <Helmet>
            <title>Test Title</title>
            <div
              // @ts-ignore
              customAttribute
            />
          </Helmet>
        );
      };

      expect(renderInvalid).toThrow(
        'Only elements types base, body, head, html, link, meta, noscript, script, style, title, Symbol(react.fragment) are allowed. Helmet does not support rendering <div> elements. Refer to our API for more information.'
      );

      global.console.error = consoleError;
    });

    it('throws on invalid strings as children', () => {
      const consoleError = global.console.error;
      global.console.error = vi.fn();

      const renderInvalid = () => {
        render(
          <Helmet>
            <title>Test Title</title>
            <link href="http://localhost/helmet" rel="canonical">
              test
            </link>
          </Helmet>
        );
      };

      expect(renderInvalid).toThrow(
        '<link /> elements are self-closing and can not contain children. Refer to our API for more information.'
      );

      global.console.error = consoleError;
    });

    it('throws on invalid children', () => {
      const consoleError = global.console.error;
      global.console.error = vi.fn();

      const renderInvalid = () => {
        render(
          <Helmet>
            <title>Test Title</title>
            <script>
              <title>Title you will never see</title>
            </script>
          </Helmet>
        );
      };

      expect(renderInvalid).toThrow(
        'Helmet expects a string as a child of <script>. Did you forget to wrap your children in braces? ( <script>{``}</script> ) Refer to our API for more information.'
      );

      global.console.error = consoleError;
    });

    it('handles undefined children', () => {
      const charSet = undefined;

      render(
        <Helmet>
          {charSet && <meta charSet={charSet} />}
          <title>Test Title</title>
        </Helmet>
      );

      expect(document.title).toBe('Test Title');
    });

    it('recognizes valid tags regardless of attribute ordering', () => {
      render(
        <Helmet>
          <meta content="Test Description" name="description" />
        </Helmet>
      );

      const existingTags = document.head.querySelectorAll(`meta[${HELMET_ATTRIBUTE}]`);
      const existingTag = existingTags[0];

      expect(existingTags).toBeDefined();
      expect(existingTags).toHaveLength(1);
      expect(existingTag).toBeInstanceOf(Element);
      expect(existingTag.getAttribute).toBeDefined();
      expect(existingTag).toHaveAttribute('name', 'description');
      expect(existingTag).toHaveAttribute('content', 'Test Description');
      expect(existingTag.outerHTML).toMatchSnapshot();
    });

    it('requestAnimationFrame works as expected', () => {
      return new Promise(resolve => {
        requestAnimationFrame(cb => {
          expect(cb).toBeDefined();
          expect(typeof cb).toBe('number');
          resolve(true);
        });
      });
    });
  });
});
