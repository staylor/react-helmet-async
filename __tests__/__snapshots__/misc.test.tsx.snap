// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`misc > API > encodes special characters 1`] = `"<meta name=\\"description\\" content=\\"This is &quot;quoted&quot; text and &amp; and '.\\" data-rh=\\"true\\">"`;

exports[`misc > API > only adds new tags and preserves tags when rendering additional Helmet instances 1`] = `"<meta name=\\"description\\" content=\\"Test description\\" data-rh=\\"true\\">"`;

exports[`misc > API > only adds new tags and preserves tags when rendering additional Helmet instances 2`] = `"<link href=\\"http://localhost/style.css\\" rel=\\"stylesheet\\" type=\\"text/css\\" data-rh=\\"true\\">"`;

exports[`misc > API > only adds new tags and preserves tags when rendering additional Helmet instances 3`] = `"<meta name=\\"description\\" content=\\"New description\\" data-rh=\\"true\\">"`;

exports[`misc > API > only adds new tags and preserves tags when rendering additional Helmet instances 4`] = `"<link href=\\"http://localhost/style2.css\\" rel=\\"stylesheet\\" type=\\"text/css\\" data-rh=\\"true\\">"`;

exports[`misc > API > only adds new tags and preserves tags when rendering additional Helmet instances 5`] = `"<meta name=\\"description\\" content=\\"Test description\\" data-rh=\\"true\\">"`;

exports[`misc > API > recognizes valid tags regardless of attribute ordering 1`] = `"<meta content=\\"Test Description\\" name=\\"description\\" data-rh=\\"true\\">"`;

exports[`misc > Declarative API > encodes special characters 1`] = `"<meta name=\\"description\\" content=\\"This is &quot;quoted&quot; text and &amp; and '.\\" data-rh=\\"true\\">"`;

exports[`misc > Declarative API > only adds new tags and preserves tags when rendering additional Helmet instances 1`] = `"<meta name=\\"description\\" content=\\"Test description\\" data-rh=\\"true\\">"`;

exports[`misc > Declarative API > only adds new tags and preserves tags when rendering additional Helmet instances 2`] = `"<link href=\\"http://localhost/style.css\\" rel=\\"stylesheet\\" type=\\"text/css\\" data-rh=\\"true\\">"`;

exports[`misc > Declarative API > only adds new tags and preserves tags when rendering additional Helmet instances 3`] = `"<meta name=\\"description\\" content=\\"New description\\" data-rh=\\"true\\">"`;

exports[`misc > Declarative API > only adds new tags and preserves tags when rendering additional Helmet instances 4`] = `"<link href=\\"http://localhost/style2.css\\" rel=\\"stylesheet\\" type=\\"text/css\\" data-rh=\\"true\\">"`;

exports[`misc > Declarative API > only adds new tags and preserves tags when rendering additional Helmet instances 5`] = `"<meta name=\\"description\\" content=\\"Test description\\" data-rh=\\"true\\">"`;

exports[`misc > Declarative API > recognizes valid tags regardless of attribute ordering 1`] = `"<meta content=\\"Test Description\\" name=\\"description\\" data-rh=\\"true\\">"`;
